### ISSUE TITLE
WPF Key Forwarding from WebView2 to Host application?

### ISSUE THREAD

**Issue by RickStrahl** [MicrosoftEdge/WebViewFeedback#468]
**Is your feature request related to a problem? Please describe.**

I'm looking to use the Web Browser control as part of an editor application that uses ACE Editor ([Markdown Monster](https://markdownmonster.west-wind.com)). Currently the application uses the Web Browser control and while it works I'm starting to run into issues with not being able to update libraries due to IE 11 support.

I've spent some time exploring WebView2 and for the most part it looks like I can duplicate the complex interop functionality with it. However, one sticking point is key forwarding:

The Web Browser control allowed me to pass through keyboard handlers for menu, toolbar and other keyboard maps in WPF. Meaning I could press say (alt-w-l) to close all documents while editing focused inside of the Web Browser control. This worked fine in the Web Browser, but has doesn't in WebView2.

AFAIK the Web Browser control would handle all locally mapped key events and if those were not handled inside of the browser, the keystrokes where passed up into the WPF host which then passed on to the appropriate WPF keyhandlers.

If that's unavailable, keymapping is going to be really difficult as essentially every possible key combination has to be monitored and handled both in the browser and in WPF.

Is there any chance that this can be added?

**Describe the solution you'd like and alternatives you've considered**

I would like this to work the same as it does in the IE Web Browser control:

* Any client side handled keyboard events are processed in the browser
* Any key events that go unhandled are passed up to WPF

**Comment by champnic** [comment]
Thanks for checking #112. I'm tracking this on our backlog separately. Anything we build for our core Win32 control will also be projected in some way for our .NET controls, so it's possible we'll come up with a solution that solves both of these issues at once, but for now we'll keep both issues. Thanks!

**Comment by RickStrahl** [comment]
Any word on this?

I've been looking to see if there's some way to work around this, and I have a partial solution of keyforwarding from the DOM but it's pretty unreliable. Having this 'just work' as it did with the Web Browser control would be a much better solution.

**Comment by champnic** [comment]
If your main use-case are accelerator keys, then we've exposed the [ICoreWebView2::add_AcceleratorKeyPressed](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2controller?view=webview2-1.0.664.37#add_acceleratorkeypressed) in the regular WPF KeyDown event of the WPF WebView2 control. It doesn't capture every keypress though.

**Comment by RickStrahl** [comment]
For me specifically it's the accellerators (Ctrl/Alt/Win/Func combos) that are important. It's possible that the old control only forwarded special key combos - never checked since the only thing I cared about was the menu/command forwarding ones.

I assume in WPF it's this:

```csharp
 private void webView_AccessKeyPressed(object sender, AccessKeyPressedEventArgs e)
        {
            Debug.WriteLine(e.Key.ToString());
        }
```

But this doesn't seem to actually ever fire. Whatever key combos I fire don't show up in the debug window (while in an edit box - or in this case in ACE editor).

**Comment by ukandrewc** [comment]
@RickStrahl Until this is resolved you can hook the keyboard directly using 'SetWindowsHookEx', etc.

**Comment by RickStrahl** [comment]
The issue isn't so much capturing the keyboard commands. There are ways to look at all keys. The problem is getting the 'right' keys forwarded to WPF to process. Figuring out what keys were already handled, and putting together 3 key accelerator pairs is not easy to do at the app level. I started down this path with some experimental code and I'm getting lots of edge cases and false positives to pass on due to multi-event tracking and timing you have to do to even think about capturing all the keys needed.

This is why the old WebBrowser control was so nice both in WPF and WinForms: It handled all that for you - only passed through what wasn't handled, and passed all those keys onto WPF in a natural way. It just worked.

For Interop between the browser and a top level host application I think this is a really important scenario and one that I would prefer not to have to implement (probably incorrectly) in every end-user application. 

Not saying this because I don't want to get my hands dirty, but because I think getting this right is really difficult.  It's a system level task that requires deep knowledge about how keys are handled and injected into the key buffers. IOW, manual implementation is very likely to get this wrong almost every time, leading to badly behaving Windows applications. I think nobody wants that including Microsoft... :smile:

**Comment by champnic** [comment]
Sorry if my comment above wasn't clear. We are raising the corresponding KeyDown events for accelerators, so you should be able to do the following:
```cs
        private void webView_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.Key == Key.P && (Keyboard.Modifiers & ModifierKeys.Control) > 0 && (Keyboard.Modifiers & ModifierKeys.Alt) > 0)
            {
                // Do something for Ctrl + Alt + P

                // Mark handled to cancel WebView2 handling
                e.Handled = true;
            }
        }
```

Let me know if that doesn't work for you, or you still have concerns about this approach. I've noticed a place in our code where we are missing hooking into the InputManager, which is potentially why AccessKeyManager isn't working here. I've opened a bug to follow-up on our side. Thanks!

**Comment by RickStrahl** [comment]
@champnic The problem is that it's not that easy as capturing a single key event. First the modifier combinations don't always come as both Alt-P down (for example) - some of it is Alt then W(indow) C(lose) for example. This doesn't capture in just the KeyDown event. You now have to capture key down track the alt key, then check for a period of time for other keys etc. 

I say this because I started down this path to try to work around this - in my version I'm doing it inside of the DOM though to check for the modifier keys there (to avoid the JS -> .NET transition until needed) and only when state changes passing the keys to .NET. It works - sort of but even so it's still unreliable with some key combos.

What I'm getting at is that this is really not easy to implement at the app level. It's doable but to do it right **takes a lot of effort**. I'm suggesting that if it's possible to do this natively inside of the control and pass this forward in the same way to old controls did, that would **avoid a lot of pain for a lot of developers**. Any application that works with input inside of the Web Browser is likely going to need this functionality to truly integrate into a desktop application.

The other issue I see here with what you suggest is that intercepting every keystroke inside of .NET is likely going to cause a bit of overhead. I'm using a sophisticated editor (ACE Editor) in JavaScript and typing speed is rather important - I suspect sending every key event into .NET and checking for special keys is going to introduce some typing latency. Handled at the control (Win32 level) this is likely to cause much less overhead than the marshalled calls into .NET I suspect.

I think I've made my point here (I'm starting to repeat myself) - I get it if you don't want to implement this but I think there's a lot of value for a lot of developers in doing so even if it's kind of a hidden feature. When you need it, self-implementing is going to waste days of dev and testing to get it right.

**Comment by champnic** [comment]
Thanks for the added context @RickStrahl. Sounds like the base case of keyboard shortcuts/accelerators are working, but we're missing support for the mnemonics for menus (alt and then w, for example). I'll update our backlog scenario to reflect that and make it more clear. Does that capture the core of your ask?

**Comment by RickStrahl** [comment]
@champnic To be clear - I'm asking for key forwarding behavior similar to the old Web Browser control which forwards everything that isn't handled inside of the browser DOM to the shell. Take a look at how the old WB control works - there it 'just works'. Control shortcuts, menus etc. The only tricky ones are the multi-key mnemonics and key chords, but ultimately I'd like to see those just forwarded to the host if not handled in the DOM.

**Comment by mediabuff** [comment]
Webview2 needs to implement IKeyboardInputSink

 https://docs.microsoft.com/en-us/dotnet/api/system.windows.interop.ikeyboardinputsink
to interoperate with wpf

https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/windows-forms-and-wpf-interoperability-input-architecture?view=netframeworkdesktop-4.8

**Comment by RickStrahl** [comment]
@mediabuff Thanks for pointing out the specific APIs that handle this. 

I really hope support for this makes it into the WebView2 control. The keyforwarding for me is a show stopper and prevents me from moving forward in the short term.

**Comment by mediabuff** [comment]
Absolutely! For many really world apps.

Also deep navigation integration - jointly with the hosting app on journal history - aka WFP/frame navigation is a must as well

**Comment by albahari** [comment]
Until 1.0.781, there was a nice workaround: you could handle the AcceleratorKeyPressed event, which exposes the WPARAM/LPARAM message properties, and then manufacture a windows message and send it to the main menu. Now that AcceleratorKeyPressed is no longer exposed, you must handle KeyDown, which doesn't expose the LPARAM property. If you know how to reconstruct the LPARAM from a KeyEventArgs, let me know.

**Comment by RickStrahl** [comment]
Ok, so here's an update on the key forwarding situation after a bunch more experimentation I think I found a solution to Menu and Toolbar shortcut forwarding issues.

After a bunch more experimentation I found that most 'special' keys are in fact forwarded into the host WPF form. Control keys and function keys seem to actually fire. 

However `alt` keys that activate menus and shortcuts **do not work** and I think the reason for that is that they are getting sent but are getting ignored by the host Window because the window doesn't have focus. Because the window doesn't have focus the Alt key operation doesn't trigger the default Windows menu and shortcut activation functionality.

So it's possible to intercept the alt key via `KeyDown` on the WebBrowser control and then explicitly fire it into the WPF window after setting focus to it:

```cs
private void WebBrowser_KeyDown(object sender, System.Windows.Input.KeyEventArgs e)
{
    // Handle Alt-Key forward to form so menus work
    if (e.Key == System.Windows.Input.Key.LeftAlt)
    {
		Model.Window.Focus();
		
		// must be out of band
                Model.Window.InvokeAsync( () => SendKeys.SendWait("%"));

                e.Handled = true;
    }
}
```
Wrote up a blog post here:

[WebView2: Forwarding Alt Keys to host WPF Window](https://weblog.west-wind.com/posts/2021/Apr/15/WebView2-Forwarding-Alt-Keys-to-a-Host-WPF-Window)


I think this resolves the issue, but it still would be nice if this could happen automatically as it did for the old IE WebBrowser control so the extra key monitoring (which adds overhead especially for edit operations) isn't necessary.

**Comment by noseratio** [comment]
This keyboard issue is hindered by the fact that the WebView2 host child window (of WPF's `HwndHost`) and the actual Chromium HTML renderer window reside in separate processes.  

So, when the focus is inside a WebView2, the WPF UI thread itself is focus-inactive, and the keyboard messages are posted to Chromium's own UI thread (in a separate process). Thus, the WPF menu accelerators and even some system hotkeys (e.g., Alt+Space for system menu) just don't work. The corresponding `WM_KEYDOWN`/`WM_KEYUP` messages simply don't reach the WPF thread's message queue. 

As a workaround, I've played a bit with [WH_KEYBOARD_LL](https://docs.microsoft.com/en-us/windows/win32/winmsg/about-hooks#wh_keyboard_ll), with not much success so far.  

I'm able to reliably intercept all the desired keystrokes from the Chromium's thread (e.f., Alt+F for Menu/File). However, I can't automatically route them through the WPF hierarchy, _without_ first setting the keyboard focus to the main window,  i.e., `Keyboard.Focus(Application.Current.MainWindow)`. This is similar to what @RickStrahl does above.

Some things I've tried to get the keyboard events routed, while the focus stays in `WebView2`:

- Calling [`ComponentDispatcher.RaiseThreadMessage()`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.interop.componentdispatcher.raisethreadmessage?view=net-5.0);
- Rising routed keyboard events with `Application.Current.MainWindow.RaiseEvent()`;
- Calling [`IKeyboardInputSink.TranslateAccelerator()`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.interop.ikeyboardinputsink.translateaccelerator?view=net-5.0). 

As to `IKeyboardInputSink.TranslateAccelerator`, I believe, it's for reverse scenarios, i.e., when the container gives a hosted control a chance to handle a keystroke. That's _conceptually_ close to `IOleInPlaceActiveObject::TranslateAccelerator` from the golden age of COM. 

While the thing I'm looking for would be _conceptually_ close to COM's `IDocHostUIHandler::TranslateAccelerator` (or `IOleInPlaceFrame::TranslateAccelerator`, or `IOleControlSite::TranslateAccelerator`). 

Logically, [`IKeyboardInputSite`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.interop.ikeyboardinputsite?view=net-5.0) would be the right place for it, but it doesn't have anything like `TranslateAccelerator`.

Maybe WPF has a proper secret solution for this scenario, but [I couldn't find one](https://source.dot.net/#q=TranslateAccelerator).  I wish `WebView2` was open-source, it might be easier to tackle this problem from that end. 

Anyhow, I'd be reluctant to convert a WPF project to use `WebView2` at this stage, until we could reliable use accelerator keys on native WPF controls. We shouldn't have to go through some sort of mumbo-jumbo workarounds for a such a basic thing like using keyboard for UI navigation.

**Comment by szanto90balazs** [comment]
@champnic I read about all the keyboard event related issues and wondering are they on the roadmap to tackle them for the near future and do you have a sneak peak into what changes likely / unlikely coming that you can share with us? Many thanks 👍 I'm particularly interested in the scenarios, when the WebView2 is the main input sink, but any unhandled DOM key events are surfaced to the WPF world. Good example would be to let the HTML/JS document add event handlers for Ctlr+SomeChar override WPF events (Ctrl+Somechar input binding).

**Comment by champnic** [comment]
@szanto90balazs We are currently in the design phase of this work, but have a good direction forwards here. We are going to change the main input sink from the WebView2 out-of-proc HWND to be the in-proc app HWND, which should enable full support for keyboarding, mouse, etc. in WPF and Winforms :) It will require some additional changes to our controls (to make sure KeyDown isn't getting fired multiple times), but should fix a ton of issues and enable new capabilities.

**Comment by szanto90balazs** [comment]
@champnic Sounds wonderful, thanks for sharing, much appreciated.

**Comment by hakzhyena** [comment]
Ok, I am late to the party but I have the same requirement for WebView2 embedded in my MFC application. I have the handle to the parent window and I want parent window to take care of keyboard shortcuts. Please consider this @champnic . 
Also, from the example in https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2controller?view=webview2-1.0.1150.38 for `add_AcceleratorKeyPressed`, what is `m_appWindow->GetAcceleratorKeyFunction()` referring to? 
Please give a similar workaround for C++ temporarily to handle keyboard shortcuts such as Ctrl + N, Ctrl + etc.

**Comment by hakzhyena** [comment]
Hi @champnic , thanks for the input. I was able to get the Ctrl key combinations working but as @RickStrahl  has mentioned above, the Alt keys work little differently. I followed his approach and following is my code in C++:
```
if (command == "IDM_ALT_PRESS")//at this point Alt key press is already captured and args->put_Handled(TRUE) is also done so that browser doesn't handle it again 
{
MainWindow()->SetFocus();
MainWindow()->PostMessage(WM_KEYDOWN, VK_MENU);
return;
}
```
Interestingly it works when I press Alt two times when WebView2 is in focus and then it toggles but for the first time it doesn't work. Any ideas?

**Comment by hakzhyena** [comment]
Hi @champnic , any thoughts on the issue above please?

**Comment by champnic** [comment]
Unfortunately no. You could open a separate GitHub issue to track that if you like. We are currently working on a change which would make input go to the host windows first before the WebView2 which would make this scenario much easier, but I don'thave a clear timeline on when that will ship yet.

**Comment by leaanthony** [comment]
> @szanto90balazs We are currently in the design phase of this work, but have a good direction forwards here. We are going to change the main input sink from the WebView2 out-of-proc HWND to be the in-proc app HWND, which should enable full support for keyboarding, mouse, etc. in WPF and Winforms :) It will require some additional changes to our controls (to make sure KeyDown isn't getting fired multiple times), but should fix a ton of issues and enable new capabilities.

Hi @champnic - just wondering where we are with this. Many thanks 🙏

**Comment by champnic** [comment]
Hey @leaanthony - It looks like the majority of the code has been implemented and is going through a code review now. This is relatively complex and large change, so may take a while before it gets checked in (ie. up to a couple weeks, not a day). I think the intent is also to have it be an opt-in behavior change, to avoid app-compat issues. Thanks!

**Comment by pareekvipul** [comment]
Hi @champnic  - We are also facing the same issue with the key down events  so can you please provide us the ETA for your fix release or prerelease.Thanks!

**Comment by champnic** [comment]
It's still in code review - I'm reaching out to the owners for updated status. Thanks!

**Comment by Eudritch** [comment]
> It's still in code review - I'm reaching out to the owners for updated status. Thanks!
> 

If he had contacted you, could you kindly provide an update?

**Comment by noseratio** [comment]
This is a cornerstone issue. It affects anyone who hosts WebView2 alongside with any native UI on Windows Desktop. 

It affects Blazor Hybrid as well: https://github.com/dotnet/maui/issues/2341.

**Comment by zhuhaichao518** [comment]
@criddell We do not handle all keys in AcceleratorKeyPressed. We only handle when Control/Alt or a key does not map to a character is down. I believe ctrl-shift-1 should work, shift-1 will not work.
The focus is currently on the browser side (you can check it out with spy++ and see details of processes). We are going to have an experimental feature on the host side.
The code is pending approval. We will be able to try it out using a feature flag in AdditionalBrowserArguments (both for Win32 and .NET).  We will be able to handle all keys by then.
@Eudritch @pareekvipul We will provide a new API to handle keys which are not handled by DOM. I will notify you when it enters into the experimental release. Then we will have both point of time to handle keys -- All before browser and after DOM.

**Comment by Eudritch** [comment]
It's been two months, is there any update?

**Comment by zhuhaichao518** [comment]
@Eudritch the experimental feature will be released in Edge 109, which can be enabled with AdditionalBrowserArguments = "--enable-features=msWebView2BrowserHitTransparent". All keys will trigger OnKeyDown in this mode.
The new API which provides key handlers after DOM is under code review. 
@RickStrahl we would like to confirm the keyforwarding sequence of the IE Web Browser control: Keys not handled by DOM may be handled by browser (e.g., Ctrl+F find bar). For IE Web Browser control, may you help confirm does browser commands like Ctrl+F happen before the key is passed up to WPF or after that? I'm trying to figure out on my side but it would be great to hear from you to be more sure.

**Comment by RickStrahl** [comment]
@zhuhaichao518 I don't have any code to look at any longer with the WebBrowser, but I do remember distinctly that the WebBrowser would not forward keys that are handled internally.

This means:

* Internal commands that the browser consumes (ctrl-f/ctrl-q)
* Keys that are handled in JS code and not bubbled up once handled in JS code

You couldn't handle **everything** in .NET and that seems to be correct behavior to me. Handled keys should not pass to .NET. 

This is limiting but if you really need that you can then use JavaScript to intercept those keys and forward them to .NET via private handling.

The scope of my original request is specifically to more advanced key combination - specifically the Alt-Menu sequences that trigger application menus, but essentially any unhandled keys should forward to .NET.

I've been able to work around everything **except for the menu mnemonics**. (Alt-W-C to close in my app). I have a hack that sort of works, but it's hideously complex (JS code + .NET code + managing key timings) and not reliable as it relies on checking keys in 6 different event handlers across JS and .NET.

**Comment by zhuhaichao518** [comment]
@RickStrahl Thanks for the information! This is quite useful for us on how to implement the new API to handle keys unhandled by browser. Hopefully this API will make the workaround easier.
I would like to mention that this API will be async as a result of chromium browser process architecture -- If you press A, B and when B is triggerred in KeyDown event, it might happen that A is still not triggered in the new API. This should also happen in your JS code + .NET code workaround (because JS code is also async with the main process).

**Comment by novac42** [comment]
@RickStrahl Regarding the new API to handle keys unhandled by browser we are working on, it will be a deferral interface to replace GetKeyState, but in our current design the keys supported are limited to SHIFT, CTRL, ALT, WIN and the toggle state of Numlock, Capslock, StrollLock. Would these keys be enough for your issue? Or you'd need to know the state of keys other than those?

**Comment by RickStrahl** [comment]
I'm not really sure what that means. 

I don't need yet another event that tells me of some obscure key state. There are already plenty of those both in JavaScript inside of the browser, and in .NET. Keys in isolation aren't very useful. 

What we are looking for here is key forwarding not capturing yet another event that you have to handle to check for specific keys. The point of this whole thread is that **coordinating all those keystates into a coherent whole** or at minimum letting keys pass through to the host that aren't handled inside of the browser (either by JavaScript or by the Browser shell).

Again, I point you at the old WebBrowser control and how it handles keys that aren't handled inside of the browser: they are passed through to the host, so complex mnemonics fire in WPF and the operating system without the control or the browser having to do anything. 

Go look at how that works to get an idea what we're talking about.

**Comment by novac42** [comment]
@RickStrahl Sorry for the confusion - let me rephrase. A common usage of a sync key event API is use it with ([GetKeyState function (winuser.h) - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeystate)) API to determine if another key(such as control, shift, a, b, c...) is also down when the key event is triggered. However, due to the browser process architecture, we will provide an async API to get keys unhandled by browser, and if you try to call [GetKeyState](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeystate) inside this API in order to get the key state of another(target) key when the key is down, it is too late because it is possible that user press/release the target key during the async time. In order to solve this problem, we will provide a new interface to substitute to this interface. What we wonder is that do we need to provide all keys (a,b,c,...) in this substitution API, or just accelerator keys are sufficient(shift,control,win,alt)?  If there is not such use case, we prefer not to provide all keys because for us, we need to scan the whole keyboard and restore all key states if we have this use case.

**Comment by Eudritch** [comment]
@novac42 Handle all key and mouse events from Windows API PostMessage - What we need is a method for creating a browser app that can handle all mouse and keyboard input from Windows PostMessage even when the browser is not active. We want to be able to simulate these inputs as if they were being performed by a user. To accomplish this, we will forward all key and mouse events to the browser app, which will respond appropriately. This would allow to build various types of browser apps with unique features.

For example, you could use a stylus to trace over an image, and the browser app below or above with a passthrough state would capture the trace, which you could then save/share. This is just one example of what you could do with it, but the possibilities are limitless!

In simpler term: We need a way to control a browser app with mouse and keyboard, even when not in use. We'll send the input to the app and it will respond like a user.

**Comment by Eudritch** [comment]
I'm not sure how the browser would handle an out-of-bounds click, or a click emitted outside of the browser rectangle, but we essentially need a way to manage all keyboard and mouse events.

**Comment by ClosetBugSlayer** [comment]
> @Eudritch the experimental feature will be released in Edge 109, which can be enabled with AdditionalBrowserArguments = "--enable-features=msWebView2BrowserHitTransparent". All keys will trigger OnKeyDown in this mode. The new API which provides key handlers after DOM is under code review.

The Edge 109 experimental feature gives me the pass-through access to open my main menu, but commands like Alt+Space and Alt+F4 do not make it through. They do get trapped by ICoreWebView2AcceleratorKeyPressedEventHandler.Invoke().

IKeyboardInputSink.TranslateAccelerator() is getting called on my HwndHost, but there is no way to forward the MSG into something like IOleInPlaceActiveObject.TranslateAccelerator() the way we could with IWebBrowser2. Therefore the WPF parent tab control is swallowing keys like the arrows and Home/End which is making it difficult to edit inside HTML forms.

**Comment by champnic** [comment]
We have an experimental feature for enabling it:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2controlleroptions.allowhostinputprocessing?view=webview2-dotnet-1.0.2730-prerelease

Alt+Space and Alt+F4 should be fixed in runtimes 128.0.2663.0+.

**Comment by drawbyperpetual** [comment]
I haven't been able to get `msWebView2BrowserHitTransparent` to work on WinUI 3. Has propagating the key events been implemented there? If not, will that be addressed with `AllowHostInputProcessing`?

**Comment by champnic** [comment]
@drawbyperpetual Sorry for the late reply - this only affects our WPF and Winforms controls.

**Comment by drawbyperpetual** [comment]
@champnic Thanks for the reply. So how can I receive key events on WinUI 3?

**Comment by segevfiner** [comment]
So it looks like `msWebView2BrowserHitTransparent ` doesn't do anything for the composition control ATM? As I'm still getting the same issues that I got without it even with it enabled when using the composition control. Such as Home/End getting caught by a parent tab control...

**Comment by DerValiser** [comment]
> We have an experimental feature for enabling it: https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2controlleroptions.allowhostinputprocessing?view=webview2-dotnet-1.0.2730-prerelease
> 
> Alt+Space and Alt+F4 should be fixed in runtimes 128.0.2663.0+.

How many further months this elementary "feature" would be experimental? 
It's now 15 months. All other points in the first patch already have been released, except AllowHostInputProcessing option.
What is the current state? Whom we have to ask in order to get it merged?

**Comment by IMFJS** [comment]
@DerValiser The most recent SDK update (1.0.3351.48) finally moved this API from preview to stable: https://learn.microsoft.com/en-us/microsoft-edge/webview2/release-notes/?tabs=winrtcsharp#allow-input-event-messages-to-pass-through-the-browser-window

As far as I can tell, nothing has changed compared to the experimental flag that we used before to work around the respective input issues. Especially it still causes this:
> I noticed some odd behavior that happens when using either the `msWebView2BrowserHitTransparent` feature flag or the `AllowHostInputProcessing` API in the 1.0.2839-prerelease: Open the Find dialog via Ctrl+F and then close it either by pressing Escape or clicking on the X. After the dialog closed, the focus won't return to the host application as one would expect. Without the feature flag or the `AllowHostInputProcessing` setting, the focus returns to the host application as expected.
> 
> I am able to reproduce this in the WebView2APISample project from the WebView2Samples repo by adding the `msWebView2BrowserHitTransparent` feature flag in the `//! [CreateCoreWebView2EnvironmentWithOptions]` section of AppWindow.cpp.

**Comment by RickStrahl** [comment]
Does anybody have an example of how to access `AllowHostInputProcessing` in a WPF application? Or is that even possible there? 

I'm not sure if the WPF and WinForms SDKs expose a controller and that's why it's not showing up on `CoreWebView2.Controller`? 

If this doesn't work in the .NET SDKs why not?

**Comment by Areopagitics** [comment]
@IMFJS does this mean it's kinda solved? how do we update?

**Comment by pushkin-** [comment]
@RickStrahl is this what you're looking for? (I'm using WinForms, but hopefully it's similar enough)

```
CoreWebView2Environment environment = await CoreWebView2Environment.CreateAsync(options: new CoreWebView2EnvironmentOptions());
CoreWebView2ControllerOptions options = environment.CreateCoreWebView2ControllerOptions();
options.AllowHostInputProcessing = true;
await webView.EnsureCoreWebView2Async(environment, options);
```

**Comment by RickStrahl** [comment]
@pushkin - Thank you, yes that worked! I couldn't find the right place to apply this.

The good news in my application ([Markdown Monster](https://markdownmonster.west-wind.com)) `options.AllowHostInputProcessing` seems to have fixed the issues I've been having with trying to handle things like menu accellerators (ie. Alt-W for Window Menu activation). And it just works now... previously I had explicit key intercept and forward code in JS that had to deal with complex focus shifts. It sort of worked but it wasn't natural Windows behavior. All that can be thrown out now (assuming there aren't major side effects).

Which is awesome. This has bugged the hell out of me for many YEARS.

Now I'm going to have to do a bunch of integration testing though to see if this breaking other key interactions... which I suspect is very likely, but on first glance it looks pretty good and is not changing the way any explicitly handled events either in JS code or in WPF are behaving. The likely issues will be things that are not handled in either place and are default and might have conflicts between the two.

I'm stoked though. This is a much needed fix and no matter what the better way this should be handled (and should be on by default IMHO - I'm enabling it in my WebView library wrapper).

**Comment by RickStrahl** [comment]
So here's one issue I'm running into: The Tab Key!

Tab key tabs out of the WebView into the next focused control and it does so before the WebView has control of the key. Since my application is an editor that's obviously not going to work. People type Tabs in an editor. I can't seem to intercept the Tab key in JavaScript even via Control or Document keydown - they never fire. 

**That's a fail!** 

Mainly because there seems to be no way to even intercept this behavior. But even if you could **I don't think Tab should change control focus here** because that's almost never appropriate for a complex control like this that has its own internal focus targets - you'd want to navigate the internal targets or if you're in some sort of edit control you don't want to jump out you'd want to jump to the next control **in the document*.


**Question** 
What exactly does this `AllowHostInputProcessing` do? 

Does it intercept very specific events and forward them? Or does do a wholesale forward of a certain group of events? I seem to be able to interface most other keys but the Tab key in particular is not even showing up in the event handler in JavaScript so that seems to suggest that this is specially handled.

IMHO, Tab should not be handled for focus change. If that's needed that could be manually handled but if it can't be intercepted we're screwed.

**Some Other Things that Surprisingly Work**

OTOH, tried a few other things I thought might be problematic like Alt - hotkeys that seems to be working: Typing key combos like Alt-I or Alt-C (to open dialogs) works as expected - so the JavaScript based handlers capture those.

**Comment by pushkin-** [comment]
Yeah this Tab thing was something I ran into too when I was playing around with the older command line arg.

I didn't test this recently, but someone on my team had offered this solution to this problem: (where the method `HandleHotkey` is something defined in our code).

The `ProcessDialogKey` -> `return false` is largely what solves this Tab problem I believe.

```
public class CustomWebView2 : WebView2()
{
     // PreProcessMessage only fires for these messages.
     const int WM_KEYDOWN = 0x100;
     const int WM_KEYUP = 0x101;
     const int WM_SYSKEYDOWN = 0x104;
     const int WM_SYSKEYUP = 0x105;
     const int WM_CHAR = 0x102;

      public override bool PreProcessMessage(ref Message msg)
      {
          if (msg.Msg == WM_CHAR) { return false; }

          bool keyDown = (msg.Msg == WM_KEYDOWN || msg.Msg == WM_SYSKEYDOWN);
          return HandleHotkey(new((Keys)msg.WParam), keyDown);
     }
}

public partial class MainForm : Form
{
     protected override bool ProcessDialogKey(Keys keyData)
     {
          return false; // Don't let the form handle the key, let webview2 do its thing
     }
}
```

Hopefully this helps you

**Comment by mikeduglas** [comment]
ESC key is the issue for me. WM_KEYDOWN(VK_ESCAPE) message is fired regardless of a context, where the user pressed the ESC key: in the input field or say in the opened dropdown list. In 1st case I want to close the window, in 2nd case I don't want because WebView2 should do its job: close the dropdown list.
I subclassed the controller's window and see that in both cases window procedure returns zero, that means the WM_KEYDOWN message was processed. But I think it should return 1 if it does nothing (it does nothing if the user presses ESC while in input field for example).

**Comment by RickStrahl** [comment]
@pushkin - thanks I've been working around it by intercepting the key in WPF and then firiing it back into the control. In my case there's JS code that runs that I can call into. Very roundabout way but it works. 

<img width="2400" height="947" alt="Image" src="https://github.com/user-attachments/assets/ac6afc03-fc27-4345-abf6-f24b9aae765b" />

The good news is that this can all be handled in the host without requiring anything in the JS/Html code.


@mikeduglas Ugh - that can really suck if you have default key/button handling on the form but that's relatively easy to work around, no? FWIW, the approach I describe above would also work for ESC - you should be able to assign a keybinding to ESC in the host and then choose to ignore it or forward to the proper behavior. Hassle yes - but a workable workaround most likely.


I sure wish there was a way to get this behavior just for the extended keys that are being passed through (ie. ALT, Ctrl etc.) Most other keys were coming through before anyway and between handling events in JS and handling them in the host you can capture everything. It's just the special behavior of accelerators and the Windows menu that's very, very difficult to capture as it requires handling sensitive timing in addition to the key combinations in different time intervals. 


--- 

I would still love to hear somebody from the WebView team explain what actually happens with this behavior.


The way I think this seems to work is that the host events are firing first with `AllowInputHostProcessing=true` where when it's off the control processes first with any events not handled in the control passing up to the host.

There is an Accelerator related event on CoreWebView, but I never got that to actually fire for anything - if that worked that likely would be a better way to go as it would capture just the edge cases

Most things can be handled just fine with the default behavior - it's mainly those special Windows behaviors I describe above  that are very, very difficult to capture and implement to behave the same. `AllowInputHostProcessing` fixes that but introduces new problems.

So far this list is:

* Tab handling - tab affects focus with no way to override from JS
* ESC passes up and can cause problems with the host
* Accelerators briefly put focus into the host, but then put it back into the editor but leave a weird hybrid state on the host where both the host and the WebView appear to be focused - I see this as the menu accelerators staying visible while the WebView2 is active

The latter is weird but not a big issue but I think that's mainly because my code manages focus and puts it back into the editor explicitly after most menu operations.

**Comment by RickStrahl** [comment]
FWIW, I've written up a blog post about the `AllowInputHostProcessing` feature and how I've used it in Markdown Monster here:

[Using the new WebView2 AllowHostInputProcessing Keyboard Mapping Feature](https://weblog.west-wind.com/posts/2025/Aug/20/Using-the-new-WebView2-AllowHostInputProcessing-Keyboard-Mapping-Feature)


For my use case this is now working as expected but it still took a bit of work to make the Tab handling work correctly.

IAC, `AllowInputHostProcessing` solves the original request in this issue from 2020.
